### PR TITLE
Cluster-autoscaler: Utils + more scale up flow

### PR DESCRIPTION
**Pull request by mwielgus**
kubernetes/kubernetes#24404

cc: @piosz @fgrzadkowski

### CODE REVIEW

File: cluster-autoscaler/cluster_autoscaler.go [hunk 1]
```diff
@@ -53,27 +49,55 @@ func main() {
...
+
+				// TODO: Checking if all nodes are present.
+
+				// Checks if scheduler tried to schedule the pods after thew newest node was added.
+				newestNode := GetNewestNode(nodes.Items)
+				if newestNode == nil {
+					glog.Errorf("No newest node")
+					continue
+				}
+				oldesSchedulingTrial := GetOldestFailedSchedulingTrail(pods)
```
Comment (fgrzadkowski) [inline cluster-autoscaler/cluster_autoscaler.go]: oldes**t**SchedulingTrial

File: cluster-autoscaler/cluster_autoscaler.go [hunk 2]
```diff
@@ -53,27 +49,55 @@ func main() {
...
+				if oldesSchedulingTrial == nil {
+					glog.Errorf("No oldest unschedueled trial: %v", err)
+					continue
+				}
+
+				if newestNode.CreationTimestamp.After(oldesSchedulingTrial.Add(-1 * time.Minute)) {
```
Comment (fgrzadkowski) [inline cluster-autoscaler/cluster_autoscaler.go]: Let's add a TODO that we should find a better way to verify that scheduler took into account newest node.

File: cluster-autoscaler/cluster_autoscaler.go [hunk 3]
```diff
@@ -53,27 +49,55 @@ func main() {
...
+					// Lets give scheduler another chance.
+					glog.V(1).Infof("One of the pods have not been tried after adding %s", newestNode.Name)
+					continue
 				}
```
Comment (fgrzadkowski) [inline cluster-autoscaler/cluster_autoscaler.go]: Maybe an info log at the end that we'd try to increase the size of the cluster?

File: cluster-autoscaler/utils.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+// List returns all unscheduled pods.
+func (unscheduledPodLister *UnscheduledPodLister) List() ([]*kube_api.Pod, error) {
+	//TODO: Extra filter based on pod condition.
+	return unscheduledPodLister.podLister.List(labels.Everything())
+}
+
+// NewUnscheduledPodLister returns a lister providing pods that failed to be scheduled.
+func NewUnscheduledPodLister(kubeClient *kube_client.Client) *UnscheduledPodLister {
+	// watch unscheduled pods
+	selector := fields.ParseSelectorOrDie("spec.nodeName==" + "" + ",status.phase!=" +
```
Comment (fgrzadkowski) [inline cluster-autoscaler/utils.go]: I think it's enough to list pods in phase `Pending`

File: cluster-autoscaler/utils.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+	podReflector.Run()
+
+	return &UnscheduledPodLister{
+		podLister: podLister,
+	}
+}
+
+// NewNodeLister builds a node lister.
+func NewNodeLister(kubeClient *kube_client.Client) *cache.StoreToNodeLister {
+	listWatcher := cache.NewListWatchFromClient(kubeClient, "nodes", kube_api.NamespaceAll, fields.Everything())
```
Comment (fgrzadkowski) [inline cluster-autoscaler/utils.go]: I think should take into account only nodes in state `Ready`. Unhealthy nodes are not interesting to us.

**Comment by fgrzadkowski** [comment]
Mostly minor things.

File: cluster-autoscaler/cluster_autoscaler.go (hunk 1 above)
Comment (mwielgus) [inline cluster-autoscaler/cluster_autoscaler.go]: Done.

File: cluster-autoscaler/cluster_autoscaler.go (hunk 2 above)
Comment (mwielgus) [inline cluster-autoscaler/cluster_autoscaler.go]: Done.

File: cluster-autoscaler/cluster_autoscaler.go (hunk 3 above)
Comment (mwielgus) [inline cluster-autoscaler/cluster_autoscaler.go]: We will increase the size of the cluster once @piosz finishes his PR. Then the log will be added.

File: cluster-autoscaler/utils.go (hunk 4 above)
Comment (mwielgus) [inline cluster-autoscaler/utils.go]: This is how k8s scheduler does it.

File: cluster-autoscaler/cluster_autoscaler.go (hunk 3 above)
Comment (fgrzadkowski) [inline cluster-autoscaler/cluster_autoscaler.go]: sgtm

File: cluster-autoscaler/utils.go (hunk 5 above)
Comment (mwielgus) [inline cluster-autoscaler/utils.go]: Done.

File: cluster-autoscaler/utils.go (hunk 4 above)
Comment (fgrzadkowski) [inline cluster-autoscaler/utils.go]: As discussed offline I believe we don't need to have the same options here, but I won't insist.

File: cluster-autoscaler/utils.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+		podLister: podLister,
+	}
+}
+
+// ReadyNodeLister lists ready nodes.
+type ReadyNodeLister struct {
+	nodeLister *cache.StoreToNodeLister
+}
+
+// List retuns ready nodes.
```
Comment (fgrzadkowski) [inline cluster-autoscaler/utils.go]: retu**r**ns

**Comment by fgrzadkowski** [comment]
One misspelling. Otherwise LGTM.

**Pull request merged**